### ISSUE TITLE
Numpy C-API `issubdtype`?

### ISSUE THREAD

**Issue by jbrockmendel** [numpy/numpy#10063]
TL;DR: Working in cython, it would be really convenient to have an implementation of `np.issubdtype` available for `cimport`.  Does this exist somewhere that I'm missing?

----
Context: pandas has [its own](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/numpy.pxd) numpy.pxd file for cimporting from numpy, best guess is this is from before `np.get_include()` was an option.  I'm trying to wean pandas off of this to just use the upstream `np.get_include()` instead.

Making this change breaks a bunch of `issubclass` [calls](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/inference.pyx#L709) of the form `issubclass(dtype.type, np.integer)`.  The exceptions that get raised/ignored are of the form

```
Exception TypeError: 'issubclass() arg 1 must be a class' in 'pandas._libs.lib.is_integer_float_array' ignored
Traceback (most recent call last):
  File "pandas/_libs/src/inference.pyx", line 620, in pandas._libs.lib.Validator.validate
    if self.is_array_typed():
  File "pandas/_libs/src/inference.pyx", line 709, in pandas._libs.lib.IntegerValidator.is_array_typed
    return issubclass(self.dtype.type, np.integer)
```

Dropping a `print(self.dtype, self.dtype.type)` just before the raising call produces `(dtype('O'), 79)`.  Looking at pandas' numpy.pxd file there is a function `_util_dtypestring` that includes

```
        if not PyDataType_HASFIELDS(child):
            t = child.type_num
            [...]
            elif t == NPY_OBJECT:      f[0] = 79 #"O"
```
So -- and this is where my wild speculation begins -- something about the pandas version of numpy.pxd is causing `issubclass` to recognize `79` as a class.

The workaround I'm using right now is replacing `issubclass(self.dtype.type, np.integer)` with `np.issubdtype(self.dtype, np.integer)`.  That appears to work, but the maintainers would insist on a pure-cython implementation.

Any thoughts?

**Comment by bashtage** [comment]
Isn't is just a question of translating 

https://github.com/numpy/numpy/blob/v1.13.0/numpy/core/numerictypes.py#L728

to C/Cython?

**Comment by njsmith** [comment]
issubdtype is implemented in python. It's also fairly messy. I'd be nervous
about trying to move it into C – it'd be very easy to break some corner
case that people are depending on. We'll probably need to rewrite it at
some point, but I'd rather wait until we're revisiting dtypes in general.

I having trouble understanding why you need to change this. get_include
isn't a replacement for numpy.pxd: you need both of them. It's not at all
clear from your post that this is even a correct diagnosis of the problem
(why exactly can't you keep doing whatever you were before?), and it'd also
be good to see some evidence that this is a bottleneck before making risky
attempts to optimize it.

**Comment by eric-wieser** [comment]
> issubdtype is implemented in python. It's also fairly messy.

It's [a lot cleaner in 1.14] - if you ignore the behavior hidden behind a deprecation, it becomes just:
```python
def issubdtype(arg1, arg2):
    if not issubclass_(arg1, generic):
        arg1 = dtype(arg1).type
    if not issubclass_(arg2, generic):
        arg2 = dtype(arg2).type
    return issubclass(arg1, arg2)
```

**Comment by jbrockmendel** [comment]
> It's not at all clear from your post that this is even a correct diagnosis of the problem
(why exactly can't you keep doing whatever you were before?)

Experience shows that me being clueless is a pretty reasonable hypothesis.

"Not broken, don't fix it" is a viable option.  The instigating factor here is an attempt to modernize setup.py and in particular start using `cythonize`.  In the status quo, extensions' `include_dirs` get both `pandas/_libs/src` (which is where numpy.pxd lives) _and_ `np.get_include()`.  Part of my implicit assumption is that this is an anti-pattern, that this could lead to ambiguity as to where things are defined.  Another implicit assumption is that `np.get_include()` is the "one obvious way to do it", or at least the one supported way.

> get_include isn't a replacement for numpy.pxd: you need both of them. 

Can you expand on this?  Is my assumption that pandas' numpy.pxd was put in place before np.get_include() was an option incorrect?  Aside from this exact issue, why are both of them needed?

> I'd be nervous about trying to move it into C – it'd be very easy to break some corner
case that people are depending on

Yah, the better scenario is probably figuring out why 79 is recognized as a class with numpy.pxd but not without.  (Even if changing the code turns out to be a non-starter, I'd be interested in understanding what's going on here.)

**Comment by jbrockmendel** [comment]
@bashtage AFAICT the issue is that without the numpy.pxd file, `np.dtype('O').type` resolves to `79`, so `issubclass` raises.  In the status quo with numpy.pxd, `np.dtype('O').type` is `<type 'numpy.object_'>`.  Something is happening in C-land that changes that behavior.

**Comment by eric-wieser** [comment]
> Dropping a print(self.dtype, self.dtype.type) just before the raising call produces (dtype('O'), 79). 

What does `type(self.dtype)` print?

**Comment by njsmith** [comment]
@jbrockmendel 

> Can you expand on this? Is my assumption that pandas' numpy.pxd was put in place before np.get_include() was an option incorrect? Aside from this exact issue, why are both of them needed?

I'm pretty sure `np.get_include` predates the existence of pandas by many years. `np.get_include` is how you get the path to numpy's C API headers, which are absolutely necessary to use numpy's C API. I'm sure pandas has always been using this. But by itself, it doesn't let you use numpy from cython. For that you need a .pxd file, to translate the C API into a Cython API.

It sounds like what's going on here is that you're switching to using the version of `numpy.pxd` that Cython includes in their source tree. Unfortunately, that version [overrides the meaning of `dtype.type` to return the type number](https://github.com/cython/cython/blob/8ccdda6dd16f6ffafef2511fdee440bd127e0f24/Cython/Includes/numpy/__init__.pxd#L168), rather than the type object like happens in regular Python (and in Cython if you don't override it at the `cdef` level, via Cython's regular handling for Python objects). Presumably pandas's version of `numpy.pxd` didn't do that.

Not sure what to tell you here. The difficulty with using Cython's `.pxd` is that if it's missing something, or does the wrong thing, then you have to submit a patch to cython and then wait for a new cython release before you can fix pandas... in any case, it sounds like this is between pandas and cython and there's nothing for numpy to do, so closing this.

**Comment by jbrockmendel** [comment]
@njsmith Thanks for walking me through this; it's a long learning curve and your time and effort is appreciated.

**Comment by eric-wieser** [comment]
> Unfortunately, that version overrides the meaning of dtype.type to return the type number, rather than the type object like happens in regular Python

This strikes me as a bug in cython

**Comment by jbrockmendel** [comment]
@eric-wieser 

> This strikes me as a bug in cython

I just opened an issue over there pointing them here.

> What does type(self.dtype) print?

I'll give it a shot and let you know.  Based on njsmith's comment it seems like for the time-being I should just make my peace with the existence of pandas' custom numpy.pxd file.
